neuralmonkey/encoders/attentive.py

```diff
@@ -0,0 +1,117 @@
+from typing import Set, cast
+
+import tensorflow as tf
+from typeguard import check_argument_types
+
+from neuralmonkey.model.stateful import TemporalStatefulWithOutput
+from neuralmonkey.model.model_part import ModelPart, FeedDict
+from neuralmonkey.nn.utils import dropout
+from neuralmonkey.dataset import Dataset
+from neuralmonkey.decorators import tensor
+from neuralmonkey.attention.base_attention import (
+    get_attention_states, get_attention_mask, Attendable)
+
+
+class AttentiveEncoder(ModelPart, TemporalStatefulWithOutput):
+    """An encoder with attention over the input and a fixed-dimension output.
+
+    Based on "A Structured Self-attentive Sentence Embedding",
+    https://arxiv.org/abs/1703.03130.
+
+    The encoder combines a sequence of vectors into a fixed-size matrix where
+    each row of the matrix is computed using a different attention head. This
+    matrix is exposed as the ``temporal_states`` property (the time dimension
+    corresponds to the different attention heads). The ``output`` property
+    provides a flattened and, optionally, projected representation of this
+    matrix.
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(self,
+                 name: str,
+                 input_sequence: Attendable,
+                 hidden_size: int,
+                 num_heads: int,
+                 output_size: int = None,
+                 state_proj_size: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        """Initialize an instance of the encoder."""
+        check_argument_types()
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        TemporalStatefulWithOutput.__init__(self)
+
+        self.input_sequence = input_sequence
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.output_size = output_size
+        self.state_proj_size = state_proj_size
+        self.dropout_keep_prob = dropout_keep_prob
+
+        if self.dropout_keep_prob <= 0.0 or self.dropout_keep_prob > 1.0:
+            raise ValueError("Dropout keep prob must be inside (0,1].")
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+    # pylint: enable=too-many-arguments
+
+    @tensor
+    def attention_states(self) -> tf.Tensor:
+        return dropout(get_attention_states(self.input_sequence),
```

Comment:
nevim jestli to je systematický nebo ne, ale většinou se snažim to dělat tak, že se dropout volá až po vrácení (čili rule of thumb je pokud jako funkce vyrábim nějakej tenzor, vracim ho v původním stavu) - tohle by totiž pak nebyly `attention_states` ale `dropped_attention_states` nebo tak něco. Taky nevím, jestli je úplně bezpečný to po dropoutu kešovat. 

zajímavá otázka taky je, jestli chceš použít dvakrát ten samej dropout jak v `attention_weights` a v `temporal_states` nebo jestli chceš aby se to pokaždý dropoutovalo jinak. jestli ten samej, tak bych to asi nechal takhle, ale napsal bych docstring nebo tak

Reply:
Přijde mi, že chceme mít dropout stejnej. Ostatně koukám, že tohle je úplně stejná situace jako v [`Attention`](https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/attention/feed_forward.py#L41) (odtud jsem to asi zkopíroval). Ale už jsem z toho udělal atribut s podtržítkem.